File: bindings/ruby/examples/streaming.rb
# frozen_string_literal: true

# Streaming inference example for the Ainos Ruby SDK.
#
# Run with:
#   bundle exec ruby examples/streaming.rb
#
# Make sure the Ainos daemon is running on 127.0.0.1:9500
# and the AINOS_TOKEN environment variable is set.

$LOAD_PATH.unshift(File.expand_path('../lib', __dir__))

require 'ainos'
require 'logger'

# Configure logging
logger = Logger.new($stdout)
logger.level = Logger::WARN

# Create a client
client = Ainos::Client.new do |config|
  config.host = ENV.fetch('AINOS_HOST', '127.0.0.1')
  config.port = ENV.fetch('AINOS_PORT', 9500).to_i
  config.token = ENV['AINOS_TOKEN']
  config.logger = logger
  config.connection_timeout = 5.0
  config.read_timeout = 120.0
end

begin
  client.connect
  puts "Connected to Ainos daemon at #{client.config.host}:#{client.config.port}"

  # Find a loaded model
  models = client.model_list
  loaded_models = models.select(&:loaded?)

  if loaded_models.empty?
    puts 'No loaded models found. Please load a model first.'
    puts "Available models: #{models.map(&:name).join(', ')}"
    exit 1
  end

  model_name = loaded_models.first.name
  puts "Using model: #{model_name}"

  # ============================================================
  # Example 1: Basic streaming with block
  # ============================================================
  puts
  puts '=' * 60
  puts 'Example 1: Basic streaming with block'
  puts '=' * 60
  puts

  print 'Response: '

  client.infer_stream(
    model: model_name,
    prompt: 'Write a haiku about Ruby programming.',
    temperature: 0.8,
    max_tokens: 128
  ) do |chunk|
    print chunk.text
    $stdout.flush
  end

  puts
  puts

  # ============================================================
  # Example 2: Enumerator-based streaming
  # ============================================================
  puts '=' * 60
  puts 'Example 2: Enumerator-based streaming'
  puts '=' * 60
  puts

  chunks = client.infer_stream(
    model: model_name,
    prompt: 'Count from 1 to 5, one number per line.',
    temperature: 0.1,
    max_tokens: 64
  )

  chunks.each_with_index do |chunk, index|
    puts "[Chunk #{index}] #{chunk.text.inspect}"
  end

  puts

  # ============================================================
  # Example 3: StreamSession with collect
  # ============================================================
  puts '=' * 60
  puts 'Example 3: StreamSession with collect'
  puts '=' * 60
  puts

  request = Ainos::InferenceRequest.new(
    model: model_name,
    prompt: 'Explain the difference between arrays and hashes in Ruby.',
    temperature: 0.7,
    max_tokens: 256
  )

  session = Ainos::StreamSession.new(client, request)
  session.each_chunk do |chunk|
    print chunk.text
    $stdout.flush
  end

  puts
  puts

  response = session.collect
  puts "Collected response: #{response.text.length} characters"
  puts "Tokens: #{response.tokens}"
  puts "Finish reason: #{response.finish_reason}"
  puts "Chunks received: #{session.chunk_count}"
  puts

  # ============================================================
  # Example 4: StreamBuilder fluent interface
  # ============================================================
  puts '=' * 60
  puts 'Example 4: StreamBuilder fluent interface'
  puts '=' * 60
  puts

  builder = Ainos::StreamBuilder.new(client)
  builder.model(model_name)
         .prompt('What are three tips for writing clean Ruby code?')
         .temperature(0.7)
         .max_tokens(256)
         .on_chunk { |chunk| print chunk.text; $stdout.flush }
         .on_finish { |text| puts "\n\n[Complete: #{text.length} chars]" }
         .on_error { |error| puts "\n[Error: #{error.message}]" }
         .start

  puts

  # ============================================================
  # Example 5: StreamAccumulator
  # ============================================================
  puts '=' * 60
  puts 'Example 5: StreamAccumulator'
  puts '=' * 60
  puts

  accumulator = Ainos::StreamAccumulator.new
  accumulator.on_chunk { |chunk| print chunk.text; $stdout.flush }
  accumulator.on_finish { |full_text| puts "\n\n[Accumulated: #{full_text.length} chars]" }

  client.infer_stream(
    model: model_name,
    prompt: 'Write a short greeting in three different languages.',
    temperature: 0.8,
    max_tokens: 128
  ) do |chunk|
    accumulator << chunk
  end

  puts
  puts "Chunks processed: #{accumulator.chunk_count}"
  puts "Elapsed time: #{accumulator.elapsed_time&.round(3)}s"
  puts "Tokens/sec: #{accumulator.tokens_per_second&.round(1)}"
  puts

  # ============================================================
  # Example 6: StreamUtils
  # ============================================================
  puts '=' * 60
  puts 'Example 6: StreamUtils with timing'
  puts '=' * 60
  puts

  client.infer_stream(
    model: model_name,
    prompt: 'What is the meaning of life?',
    temperature: 0.9,
    max_tokens: 64
  ) do |chunk|
    Ainos::StreamUtils.with_timing([chunk]) do |c, delta|
      puts "[+#{format('%.3f', delta)}s] #{c.text}"
    end
  end

  puts

  # ============================================================
  # Example 7: Multiple concurrent streams
  # ============================================================
  puts '=' * 60
  puts 'Example 7: Sequential streams (concurrent requires threads)'
  puts '=' * 60
  puts

  prompts = [
    'Say "Hello" in French.',
    'Say "Hello" in German.',
    'Say "Hello" in Spanish.'
  ]

  prompts.each_with_index do |prompt, i|
    print "Stream #{i + 1}: "

    client.infer_stream(
      model: model_name,
      prompt: prompt,
      temperature: 0.5,
      max_tokens: 32
    ) do |chunk|
      print chunk.text
      $stdout.flush
    end

    puts
  end

rescue Ainos::ConnectionError => e
  puts "Connection error: #{e.message}"
  puts 'Make sure the Ainos daemon is running.'
rescue Ainos::Error => e
  puts "Ainos error: #{e.detailed_message}"
rescue StandardError => e
  puts "Unexpected error: #{e.class}: #{e.message}"
  puts e.backtrace.first(3).join("\n")
ensure
  client&.disconnect
  puts
  puts 'Disconnected.'
end

File: bindings/python/examples/streaming.py
"""
Ainos SDK - Streaming Inference Example
========================================

Demonstrates streaming inference with the Ainos daemon, including:
- Basic streaming with async for loop
- Collecting all tokens into a string
- Using StreamAccumulator for progress tracking
- Handling stream cancellation
- Measuring streaming performance

Prerequisites:
    - The Ainos daemon must be running on localhost:9500
    - At least one model must be loaded

Usage:
    python examples/streaming.py
"""

from __future__ import annotations

import asyncio
import logging
import os
import sys
import time
from typing import Any

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from ainos import AinosClient, StreamAccumulator, setup_logging
from ainos.errors import AinosError
from ainos.types import InferenceChunk, ModelInfo


async def main() -> None:
    """Run the streaming example."""
    setup_logging(level=logging.WARNING)

    auth_token: str | None = os.environ.get("AINOS_AUTH_TOKEN")

    print("=" * 60)
    print("Ainos SDK - Streaming Inference Example")
    print("=" * 60)

    # ------------------------------------------------------------------
    # 1. Connect and find a model
    # ------------------------------------------------------------------
    print("\n[1] Connecting to daemon and finding a loaded model...")

    client: AinosClient = AinosClient(
        host="127.0.0.1",
        port=9500,
        auth_token=auth_token,
    )
    await client.connect()

    models: list[ModelInfo] = await client.model_list()
    loaded: list[ModelInfo] = [m for m in models if m.status == "loaded"]

    if not loaded:
        print("    No loaded models found. Please load a model first.")
        print("    Example: python examples/model_management.py")
        await client.disconnect()
        return

    model_name: str = loaded[0].id
    print(f"    Using model: {model_name}")

    # ------------------------------------------------------------------
    # 2. Basic streaming (async for loop)
    # ------------------------------------------------------------------
    print("\n[2] Basic streaming inference...")
    print("    Generating text, streaming token by token:\n")

    prompt: str = "Write a short paragraph about the future of artificial intelligence."

    stream = client.infer_stream(
        model=model_name,
        prompt=prompt,
        max_tokens=200,
        temperature=0.7,
    )

    start_time: float = time.monotonic()
    token_count: int = 0
    full_text: str = ""

    async for chunk in stream:
        print(chunk.token, end="", flush=True)
        full_text += chunk.token
        token_count += 1

        if chunk.final:
            print()
            print(f"\n    [Stream finished: {chunk.finish_reason}]")

    elapsed: float = time.monotonic() - start_time
    print(f"\n    Tokens: {token_count} in {elapsed:.2f}s "
          f"({token_count / elapsed:.1f} tokens/s)")

    # ------------------------------------------------------------------
    # 3. Using StreamAccumulator
    # ------------------------------------------------------------------
    print("\n[3] Streaming with StreamAccumulator (collect + stats)...")

    accumulator: StreamAccumulator = StreamAccumulator()

    async for chunk in client.infer_stream(
        model=model_name,
        prompt="List 3 interesting facts about the Python programming language.",
        max_tokens=150,
        temperature=0.8,
    ):
        accumulator.add(chunk)
        # Still print tokens as they arrive
        print(chunk.token, end="", flush=True)

    print("\n")
    print(f"    Accumulated text length: {len(accumulator.text)} chars")
    print(f"    Token count: {accumulator.token_count}")
    print(f"    Finish reason: {accumulator.finish_reason}")
    print(f"    Elapsed: {accumulator.elapsed:.2f}s")
    if accumulator.usage:
        print(f"    Usage: {accumulator.usage.total_tokens} total tokens")

    # ------------------------------------------------------------------
    # 4. Streaming with different parameters
    # ------------------------------------------------------------------
    print("\n[4] Streaming with different parameters (higher temperature)...")

    prompt = "Write a creative tagline for a robot barista."

    async for chunk in client.infer_stream(
        model=model_name,
        prompt=prompt,
        max_tokens=30,
        temperature=1.5,  # More creative / random
        top_p=0.9,
    ):
        print(chunk.token, end="", flush=True)

    print("\n")

    # ------------------------------------------------------------------
    # 5. Collect all tokens at once
    # ------------------------------------------------------------------
    print("\n[5] Collecting all tokens with collect()...")

    stream = client.infer_stream(
        model=model_name,
        prompt="What is 2 + 2? Answer briefly.",
        max_tokens=20,
    )
    collected: str = await stream.collect()
    print(f"    Collected text: {collected.strip()}")

    # ------------------------------------------------------------------
    # 6. Streaming with stop sequences
    # ------------------------------------------------------------------
    print("\n[6] Streaming with stop sequences...")

    async for chunk in client.infer_stream(
        model=model_name,
        prompt="Count from 1 to 5:",
        max_tokens=50,
        stop=["3"],  # Will stop at "3"
    ):
        print(chunk.token, end="", flush=True)

    print("\n    (Stopped at '3' due to stop sequence)")

    # ------------------------------------------------------------------
    # 7. Clean up
    # ------------------------------------------------------------------
    print("\n[7] Disconnecting...")
    await client.disconnect()
    print("    Done.")

    print("\n" + "=" * 60)
    print("Streaming example completed!")
    print("=" * 60)


if __name__ == "__main__":
    asyncio.run(main())